### PR TITLE
[GUI] Support drawing text on canvas

### LINKED ISSUE #1045: [GUI] Draw text on canvas
(This is an issue for GAMES 201 TAs.)

**Concisely describe the proposed feature**

Add method `text(context: str, pos: tuple, font_size: int)` to the [GUI class](https://github.com/taichi-dev/taichi/blob/6c02c401a60625f730aedd3c675974fc8efae1cd/python/taichi/misc/gui.py#L5) in Python.

**Describe the solution you'd like (if any)**
In C++ this is already implemented here:
https://github.com/taichi-dev/taichi/blob/6c02c401a60625f730aedd3c675974fc8efae1cd/taichi/gui/gui.h#L388-L396

Please bind the C++ method to Python using pybind11 here:
https://github.com/taichi-dev/taichi/blob/6c02c401a60625f730aedd3c675974fc8efae1cd/taichi/python/export_visual.cpp#L39

(Please see how `GUI.circle` is implemented.)

You may also need the font file (https://github.com/yuanming-hu/taichi_assets/tree/master/fonts/go).

In the end, make sure the font file is shipped with Taichi.

### PR DESCRIPTION
**Pull request by archibate**
Related issue = close #1045

[[Click here for the format server]](http://kun.csail.mit.edu:31415/)

### CODE REVIEW

**Comment by archibate** [comment]
Btw, why is our font resolution so bad?
图片

**Comment by archibate** [comment]
Can't we just use the platform provided API to draw text? (e.g. `TextOut` in win32)

**Comment by yuanming-hu** [comment]
Thanks! Brain fried today but I'll take a look first thing tomorrow.

File: python/taichi/misc/gui.py [hunk 1]
```diff
@@ -133,6 +133,14 @@ def rect(self, topleft, bottomright, radius=1, color=0xFFFFFF):
         self.line(c, d, radius, color)
         self.line(d, a, radius, color)
 
+    def text(self, context, pos, font_size=15, color=0xFFFFFF):
```
Comment (yuanming-hu) [inline python/taichi/misc/gui.py:136]: `color` is unused?

File: python/taichi/misc/gui.py [hunk 2]
```diff
@@ -133,6 +133,14 @@ def rect(self, topleft, bottomright, radius=1, color=0xFFFFFF):
...
+        import taichi as ti
+        # TODO: refactor Canvas::text
+        font_size = float(font_size)
+        pos = ti.vec(*pos)
+        color = ti.vec(0, 255, 0, 0)
```
Comment (yuanming-hu) [inline python/taichi/misc/gui.py:141]: Note that your canvas uses `f32` and `color` should be `(0, 1, 0, 0)`. You may also want to change it to a value dependent on your `color` argument. After changing 255 to 1 I get anti-aliased font:

Screenshot from 2020-05-29 07-44-59

**Comment by archibate** [comment]
Can't we just load the system default font, or compile the font into libtaichi.so? It's tricky to copytree/rmtree nested-dir like `external/fonts`.

**Comment by archibate** [comment]
Problem:
When in `release_mode`, we `ttf_path = "/fonts/Go.ttf"`.
When in `dev_mode`, we `ttf_path = "/external/fonts/Go.ttf"`.
So how do I detect if it's dev_mode in C++?

**Comment by archibate** [comment]
Now I just use `external/fonts` to keep consistency in release_mode, sleeping now, ff2 mt oyo fuat :)

**Comment by archibate** [comment]
How to test in release mode?

**Comment by yuanming-hu** [comment]
Thanks for implementing everything here! 

> So how do I detect if it's dev_mode in C++?

Please use `is_release()` in C++.


> How to test in release mode?

Usually I do

```bash
cd python
python3 build.py test
```

**Comment by archibate** [comment]
```
(aug) [bate@archit taichi]$ cd python
(aug) [bate@archit python]$ python build.py test
[Taichi] mode=development
[Taichi] preparing sandbox at /tmp/taichi-krj62zja
[Taichi] sandbox prepared
[Taichi] <dev mode>, supported archs: [cpu, cuda, opengl], commit 98138bdb, python 3.8.2
Traceback (most recent call last):
  File "build.py", line 26, in <module>
    assert False, "Missing environment variable PYPI_PWD"
AssertionError: Missing environment variable PYPI_PWD
```
Why it needs PYPI_PWD? Wasn't it just `test`?

**Comment by yuanming-hu** [comment]
Oh, my bad. Please use any value for that environment variable for now. In the future let's fix this on `test` mode.

**Comment by archibate** [comment]
FYI it also needs `CXX=clang++` env.
```
(guitext) [bate@archit test_env]$ p keyboard.py 
[Taichi] mode=release
[Taichi] version 0.6.6, supported archs: [cpu, cuda, opengl], commit 5cab3f28, python 3.8.2
Traceback (most recent call last):
  File "keyboard.py", line 26, in <module>
    gui.text(f'({x:.3}, {y:.3})', (x, y))
AttributeError: 'GUI' object has no attribute 'text'
```
`commit 5cab3f28`? So `build.py test` only work for the latest `[release]`?

**Comment by yuanming-hu** [comment]
Maybe you have an existing v0.6.6 installation via pip?

**Comment by archibate** [comment]
Fixed. Now comes to:
```
Traceback (most recent call last):
  File "keyboard.py", line 26, in <module>
    gui.text(f'({x:.3}, {y:.3})', (x, y))
  File "/home/bate/.local/lib/python3.8/site-packages/taichi/misc/gui.py", line 143, in text
    self.canvas.text(context, pos, font_size, color)
RuntimeError: [image_buffer.cpp:write_text@120] Font file not found: /home/bate/.taichi//external/fonts/NotoSansMono-Regular.ttf
```
`/home/bate/.taichi/`?
I want `/home/bate/.local/lib/python/site-packages/taichi/`.

**Comment by yuanming-hu** [comment]
`libdevice_path` might be a good reference.

Btw, we should avoid committing huge binary files in the repo. Could you make the font file a submodule? Also how about sticking to the font file in `taichi_assets`, [which I'm sure has a friendly license](https://github.com/yuanming-hu/taichi_assets/tree/master/fonts/go)?

Thanks for working on this!

**Comment by archibate** [comment]
Can we remove or rename https://github.com/yuanming-hu/taichi_assets now? It's legacy and no longer used except for it's release page (CI resources) anyway.

**Comment by archibate** [comment]
Working very well in release mode!
图片

File: python/taichi/misc/gui.py [hunk 3]
```diff
@@ -133,6 +133,15 @@ def rect(self, topleft, bottomright, radius=1, color=0xFFFFFF):
         self.line(c, d, radius, color)
         self.line(d, a, radius, color)
 
+    def text(self, context, pos, font_size=15, color=0xFFFFFF):
+        import taichi as ti
+        # TODO: refactor Canvas::text
+        font_size = float(font_size)
+        pos = ti.vec(*pos)
+        r, g, b = color & 0xff, (color >> 8) & 0xff, (color >> 12) & 0xff
```
Comment (yuanming-hu) [inline python/taichi/misc/gui.py:141]:
```suggestion
        r, g, b = (color >> 16) & 0xff, (color >> 8) & 0xff, color & 0xff
```

File: docs/gui.rst [hunk 4]
```diff
@@ -150,6 +150,16 @@ Paint on a window
...
 
 
+.. function:: gui.text(context, pos, font_size = 15, color = 0xFFFFFF)
+
+    :parameter gui: (GUI) the window object
+    :parameter pow: (tuple of 2) the top-left point position of the fonts / texts
+    :parameter font_size: (optional, scalar) the size of font (in height)
+    :parameter color: (optional, RGB hex) the foreground color of text
+
+    Draw a line of text on screen.
```
Comment (yuanming-hu) [inline docs/gui.rst:160]:
```suggestion
.. function:: gui.text(content, font_size = 15, color = 0xFFFFFF)

    :parameter gui: (GUI) the GUI object
    :parameter content: (``str``) the text to print
    :parameter pos: (tuple of 2) the top-left anchor position of the text
    :parameter font_size: (optional, scalar) font size (height, in pixels)
    :parameter color: (optional, RGB hex) text foreground color

    Draw text on the screen.
```

File: python/taichi/misc/gui.py [hunk 5]
```diff
@@ -133,6 +133,15 @@ def rect(self, topleft, bottomright, radius=1, color=0xFFFFFF):
         self.line(c, d, radius, color)
         self.line(d, a, radius, color)
 
+    def text(self, context, pos, font_size=15, color=0xFFFFFF):
```
Comment (yuanming-hu) [inline python/taichi/misc/gui.py:136]:
```suggestion
    def text(self, content, pos, font_size=15, color=0xFFFFFF):
```

File: python/taichi/misc/gui.py [hunk 6]
```diff
@@ -133,6 +133,15 @@ def rect(self, topleft, bottomright, radius=1, color=0xFFFFFF):
...
+        import taichi as ti
+        # TODO: refactor Canvas::text
+        font_size = float(font_size)
+        pos = ti.vec(*pos)
+        r, g, b = (color >> 16) & 0xff, (color >> 8) & 0xff, color & 0xff
+        color = ti.vec(r / 255, g / 255, b / 255, 1)
+        self.canvas.text(context, pos, font_size, color)
```
Comment (yuanming-hu) [inline python/taichi/misc/gui.py:143]:
```suggestion
        self.canvas.text(content, pos, font_size, color)
```

File: python/taichi/misc/gui.py (hunk 5 above)
Comment (yuanming-hu) [reply python/taichi/misc/gui.py:136]: Note that `content` is different from `context`.

**Pull request merged**